corehq/db.py

```diff
@@ -1,13 +1,92 @@
 from django.conf import settings
-from sqlalchemy.engine import create_engine
+import sqlalchemy
 from sqlalchemy.orm.scoping import scoped_session
 from sqlalchemy.orm.session import sessionmaker
 from django.core import signals
 
-_engine = create_engine(settings.SQL_REPORTING_DATABASE_URL)
-_session_factory = sessionmaker(bind=_engine)
+DEFAULT_ENGINE_ID = 'default'
 
-Session = scoped_session(_session_factory)
+
+def create_engine(connection_string=None):
+    # todo: this function is just a proxy for the sqlalchemy version and should be removed
+    connection_string = connection_string or settings.SQL_REPORTING_DATABASE_URL
+    return sqlalchemy.create_engine(connection_string)
+
+
+class SessionFactory(object):
+    """
+    Shim class for a single connection/session factory
+    """
+
+    def __init__(self, connection_string):
+        self.engine = create_engine(connection_string)
+        self._session_factory = sessionmaker(bind=self.engine)
+        # Session is the actual constructor object
+        self.Session = scoped_session(self._session_factory)
+
+
+class ConnectionManager(object):
+    """
+    Object for dealing with sqlalchemy engines and sessions.
+    """
+
+    def __init__(self):
+        self._session_factories = {}
+
+    def _get_or_create_factory(self, engine_id):
+        if engine_id not in self._session_factories:
+            self._session_factories[engine_id] = SessionFactory(self._get_connection_string(engine_id))
+        return self._session_factories[engine_id]
+
+    def get_session_factory(self, engine_id=DEFAULT_ENGINE_ID):
+        """
+        Returns the SessionFactory object associated with this engine id
+        """
+        # making this separate from _get_or_create in case we ever want to fail differently here
+        return self._get_or_create_factory(engine_id)
+
+    def get_scoped_session(self, engine_id=DEFAULT_ENGINE_ID):
+        """
+        This returns a class that can be instantiated and will have a session scoped
+        to the local thread.
```

Comment:
After reading this: http://docs.sqlalchemy.org/en/rel_1_0/orm/contextual.html#implicit-method-access I think you could just say that this returns a scoped session object.